core/src/main/scala/com/rallyhealth/vapors/core/data/FactTypeSet.scala

```diff
@@ -1,71 +1,72 @@
-      // Justification: This checks equality of the FactType at runtime after safely casting the value
-      fact.asInstanceOf[Fact[B]]
-  }
-
-  // TODO: Write unit tests for this to determine the expected utility of this method
-  def subset[B](implicit tb: TypeTag[B]): Option[FactTypeSet[B]] = {
-    val matchingFactTypes = types.toList.collect {
-      // TODO: Should this use <:< instead of =:= ?
-      case (_, factType) if factType.tt.tpe =:= tb.tpe =>
-        // Justification: This checks equality of the FactType at runtime after safely casting the value
-        factType.asInstanceOf[FactType[B]]
-    }
-    FactTypeSet.fromList(matchingFactTypes)
-  }
-
-  // TODO: Why is this Match implemented differently than matchAs?
-  //       Either we should remove it or rewrite the above code to use it.
-  final object Match {
+  def collector: PartialFunction[AbstractFact, Fact[A]] = castPF
 
-    def apply(fact: Fact[_]): Option[Fact[A]] = {
-      types.get(fact.typeInfo.name).flatMap(_.cast(fact))
-    }
-
-    def unapply(fact: Fact[_]): Option[Fact[A]] = apply(fact)
-  }
 }
 
 object FactTypeSet {
 
-  // TODO: Why have both empty and of methods? Shouldn't this be a NonEmptyMap?
-  def empty[A]: FactTypeSet[A] = new FactTypeSet(Map())
-
   def of[A](
     one: FactType[A],
     others: FactType[A]*,
   ): FactTypeSet[A] = {
-    new FactTypeSet(Map(one.fullName -> one) ++ others.map(a => (a.fullName, a)))
+    import cats.instances.string._
+    new FactTypeSet(NonEmptyMap.of(one.fullName -> one, others.map(a => (a.fullName, a)): _*))
   }
 
-  def fromSet[A](types: Set[FactType[A]]): Option[FactTypeSet[A]] = {
-    fromList(types.toList)
-  }
+  def fromFactsNel[A](facts: NonEmptyList[Fact[A]]): FactTypeSet[A] = fromNel(facts.map(_.typeInfo))
 
-  def fromList[A](types: List[FactType[A]]): Option[FactTypeSet[A]] = {
-    val typeMap = types.groupBy(_.fullName).collect {
-      case (name, tpe :: Nil) => (name, tpe)
+  def fromNel[A](types: NonEmptyList[FactType[A]]): FactTypeSet[A] = of(types.head, types.tail: _*)
+
+  def fromNes[A](types: NonEmptySet[FactType[A]]): FactTypeSet[A] = of(types.head, types.tail.toSeq: _*)
+
+  def validateSet[A](types: Set[FactType[A]]): Either[String, FactTypeSet[A]] = validateList(types.toList)
+
+  def validateList[A](types: List[FactType[A]]): Either[String, FactTypeSet[A]] = {
+    import cats.instances.string._
+    val m = types.groupBy(_.fullName)
+    val (duplicates, uniqueTypes) = m.partitionMap {
+      case (name, one :: Nil) => Right((name, one))
+      case (name, tooMany) => Left((name, tooMany))
     }
-    if (typeMap.size != types.size) None
-    else Some(new FactTypeSet(typeMap))
+    NonEmptyMap
+      .fromMap(immutable.SortedMap.from(uniqueTypes))
+      .map { typeMap =>
+        Right(new FactTypeSet(typeMap))
+      }
+      .getOrElse {
+        Left(
+          s"types list contains duplicate fact types: ${duplicates.map(_._1).mkString("'", "', '", "'")}",
```

Comment:
We're not checking that duplicates is non-empty before creating a `Right()`.

This line will only be called if the list is empty (and not if there are duplicates).

Reply:
Good catch! Fixed.